pallets/phala/src/mining.rs

```diff
@@ -531,28 +553,102 @@ pub mod pallet {
+		fn minimal_stake(&self, p: u32) -> BalanceOf<T> {
+			let p = FixedPoint::from_num(p);
+			let k = FixedPoint::from_bits(self.params.k);
+			let min_stake = k * p.sqrt();
+			FixedPointConvert::from_fixed(&min_stake)
+		}
+
+		/// Calcuates the initial Ve
+		fn ve(&self, s: BalanceOf<T>, p: u32, confidence_level: u8) -> FixedPoint {
+			let f1 = FixedPoint::from_num(1);
+			let score = Self::confidence_score(confidence_level);
+			let re = FixedPoint::from_bits(self.params.re);
+			let tweaked_re = (re - f1) * score + f1;
+			let s = s.to_fixed();
+			let c = self.rig_cost(p);
+			tweaked_re * (s + c)
+		}
+
+		/// Gets the max v in fixed point
+		fn v_max(&self) -> FixedPoint {
+			FixedPoint::from_bits(self.params.v_max)
+		}
+
+		/// Gets the estimated rig costs in PHA
+		fn rig_cost(&self, p: u32) -> FixedPoint {
+			let cost_k = FixedPoint::from_bits(self.params.rig_k);
+			let cost_b = FixedPoint::from_bits(self.params.rig_b);
+			let pha_rate = FixedPoint::from_bits(self.params.pha_rate);
+			let p = FixedPoint::from_num(p);
+			(cost_k * p + cost_b) / pha_rate
+		}
+
+		/// Gets the operating cost per sec
+		#[cfg(test)]
+		fn op_cost(&self, p: u32) -> FixedPoint {
+			let cost_k = FixedPoint::from_bits(self.params.cost_k);
+			let cost_b = FixedPoint::from_bits(self.params.cost_b);
+			let pha_rate = FixedPoint::from_bits(self.params.pha_rate);
+			let p = FixedPoint::from_num(p);
+			(cost_k * p + cost_b) / pha_rate
+		}
+
+		/// Converts confidence level to score
+		fn confidence_score(confidence_level: u8) -> FixedPoint {
+			const SCORES_BASE10: [u32; 5] = [10, 10, 10, 8, 7];
+			let score_base10 = if 1 <= confidence_level && confidence_level <= 5 {
+				SCORES_BASE10[confidence_level as usize - 1]
+			} else {
+				10u32
+			};
+			FixedPoint::from_num(score_base10) / 10
+		}
+	}
+
 	#[pallet::genesis_config]
 	pub struct GenesisConfig {
 		pub tokenomic_parameters: TokenomicParams,
 	}
 
 	#[cfg(feature = "std")]
 	impl Default for GenesisConfig {
+		/// Default tokenoic parameters for Phala
 		fn default() -> Self {
-			use substrate_fixed::types::U64F64 as FixedPoint;
-
-			pub fn fp(n: u64) -> FixedPoint {
+			fn fp(n: u64) -> FixedPoint {
 				FixedPoint::from_num(n)
 			}
-
 			let pha_rate = fp(1);
 			let rho = fp(100000099985) / 100000000000; // hourly: 1.00020,  1.0002 ** (1/300)
 			let slash_rate = fp(1) / 1000 / 300; // hourly rate: 0.001, convert to per-block rate
-			let budget_per_sec = fp(1000);
+			let budget_per_sec = fp(720000) / 24 / 3600;
 			let v_max = fp(30000);
-			let cost_k = fp(287) / 10000 / 300; // hourly: 0.0287
-			let cost_b = fp(15) / 300; // hourly : 15
+			let cost_k = fp(415625) / 10_000000 / 3600 / 24 / 365; // annual 0.0415625, convert to per sec
+			let cost_b = fp(88_59375) / 100000 / 3600 / 24 / 365; // annual 88.59375, convert to per sec
 			let heartbeat_window = 10; // 10 blocks
+			let rig_k = fp(3) / 10;
+			let rig_b = fp(0);
+			let re = fp(15) / 10;
+			let k = fp(100);
 
```

Comment:
Now that we have `fixed-macro` it's time to drop the `fn fp` now?

Reply:
Fixed. Got better accuracy